presto-main/src/test/java/com/facebook/presto/server/remotetask/TestHttpRemoteTask.java

```diff
@@ -0,0 +1,237 @@
+
+public class TestHttpRemoteTask
+{
+    private final CompletableFuture<Void> testComplete = new CompletableFuture<>();
+
+    @Test
+    public void testRemoteTaskMismatch()
+            throws InterruptedException, ExecutionException
+    {
+        Duration idleTimeout = new Duration(3, SECONDS);
+        Duration failTimeout = new Duration(20, SECONDS);
+
+        JsonCodec<TaskStatus> taskStatusCodec = JsonCodec.jsonCodec(TaskStatus.class);
+        JsonCodec<TaskInfo> taskInfoCodec = JsonCodec.jsonCodec(TaskInfo.class);
+        TaskManagerConfig taskManagerConfig = new TaskManagerConfig();
+
+        // Shorten status refresh wait and info update interval so that we can have a shorter test timeout
+        taskManagerConfig.setStatusRefreshMaxWait(new Duration(idleTimeout.roundTo(MILLISECONDS) / 100, MILLISECONDS));
+        taskManagerConfig.setInfoUpdateInterval(new Duration(idleTimeout.roundTo(MILLISECONDS) / 10, MILLISECONDS));
+
+        AtomicLong lastActivity = new AtomicLong();
+        HttpProcessor httpProcessor = new HttpProcessor(taskStatusCodec, taskInfoCodec, lastActivity);
+        TestingHttpClient testingHttpClient = new TestingHttpClient(httpProcessor);
+
+        HttpRemoteTaskFactory httpRemoteTaskFactory = new HttpRemoteTaskFactory(
+                new QueryManagerConfig(),
+                taskManagerConfig,
+                testingHttpClient,
+                new TestSqlTaskManager.MockLocationFactory(),
+                taskStatusCodec,
+                taskInfoCodec,
+                JsonCodec.jsonCodec(TaskUpdateRequest.class),
+                new RemoteTaskStats());
+        RemoteTask remoteTask = httpRemoteTaskFactory.createRemoteTask(
+                TEST_SESSION,
+                new TaskId("test", 1, 2),
+                new PrestoNode("node-id", URI.create("http://192.0.1.2"), new NodeVersion("version"), false),
+                TaskTestUtils.PLAN_FRAGMENT,
+                ImmutableMultimap.of(),
+                OutputBuffers.createInitialEmptyOutputBuffers(OutputBuffers.BufferType.BROADCAST),
+                new NodeTaskMap.PartitionedSplitCountTracker(i -> {
+                }),
+                true);
+
+        httpProcessor.setInitialTaskInfo(remoteTask.getTaskInfo());
+        remoteTask.start();
+        asyncRun(
+                idleTimeout,
+                failTimeout,
+                lastActivity,
+                () -> testComplete.complete(null),
+                (message, cause) -> {
+                    httpRemoteTaskFactory.stop();
+                    testComplete.completeExceptionally(new AssertionError(message, cause));
+                });
+
+        testComplete.get();
+
+        assertTrue(remoteTask.getTaskStatus().getState().isDone(), format("TaskStatus is not in a done state: %s", remoteTask.getTaskStatus()));
+        assertTrue(remoteTask.getTaskInfo().getTaskStatus().getState().isDone(), format("TaskInfo is not in a done state: %s", remoteTask.getTaskInfo()));
```

Comment:
For `TaskInfo`, this assert will be hard because my `TestingHttpClient.Processor` does not actually do JSON deserialization.